aptible/log_drain.go

```diff
@@ -0,0 +1,135 @@
+	DrainApps              bool
+	DrainDatabases         bool
+	DrainEphemeralSessions bool
+	DrainProxies           bool
+	DrainHost              strfmt.URI
+	DrainPassword          string
+	DrainPort              int64
+	DrainType              *string
+	LoggingToken           string
+	URL                    strfmt.URI
+}
+
+func (c *Client) CreateLogDrain(handle string, accountID int64, attrs *LogDrainCreateAttrs) (*LogDrain, error) {
+	request := &models.AppRequest16{
+		Database:               attrs.Database,
+		DatabaseID:             attrs.DatabaseID,
+		DrainApps:              attrs.DrainApps,
+		DrainDatabases:         attrs.DrainDatabases,
+		DrainEphemeralSessions: attrs.DrainEphemeralSessions,
+		DrainHost:              attrs.DrainHost,
+		DrainPassword:          attrs.DrainPassword,
+		DrainPort:              attrs.DrainPort,
+		DrainType:              attrs.DrainType,
+		Handle:                 &handle,
+		LoggingToken:           attrs.LoggingToken,
+		URL:                    attrs.URL,
+	}
+	logDrain := &LogDrain{}
+	params := operations.NewPostAccountsAccountIDLogDrainsParams().WithAccountID(accountID).WithAppRequest(request)
+	response, err := c.Client.Operations.PostAccountsAccountIDLogDrains(params, c.Token)
+	if err != nil {
+		return logDrain, err
+	}
+
+	if response.Payload.ID == nil {
+		return logDrain, fmt.Errorf("log drain ID is a nil pointer")
+	}
+	logDrain.ID = *response.Payload.ID
+	logDrain.Handle = *response.Payload.Handle
+	logDrain.DrainType = *response.Payload.DrainType
+	logDrain.DrainHost = *response.Payload.DrainHost
+	logDrain.DrainPort = *response.Payload.DrainPort
+	logDrain.LoggingToken = *response.Payload.LoggingToken
+	logDrain.URL = *response.Payload.URL
+	logDrain.DrainApps = *response.Payload.DrainApps
+	logDrain.DrainDatabases = *response.Payload.DrainDatabases
+	logDrain.DrainEphemeralSessions = *response.Payload.DrainEphemeralSessions
+	logDrain.DrainProxies = *response.Payload.DrainProxies
+	logDrain.Deleted = false
+
+	return logDrain, nil
+}
+
+func (c *Client) GetLogDrain(logDrainID int64) (*LogDrain, error) {
+	logDrain := &LogDrain{
+		Deleted: false,
+	}
+
+	params := operations.NewGetLogDrainsIDParams().WithID(logDrainID)
+	response, err := c.Client.Operations.GetLogDrainsID(params, c.Token)
+	if err != nil {
+		errStruct := err.(*operations.GetLogDrainsIDDefault)
+		switch errStruct.Code() {
+		case 404:
+			// If deleted == true, then the endpoint needs to be removed from Terraform.
+			logDrain.Deleted = true
+			return logDrain, nil
+		case 401:
+			e := fmt.Errorf("make sure you have the correct auth token")
+			return logDrain, e
+		default:
+			e := fmt.Errorf("there was an error when completing the request to get the app \n[ERROR] -%s", err)
+			return logDrain, e
+		}
+	}
+
+	if response.Payload.ID == nil {
+		return logDrain, fmt.Errorf("payload.ID is a nil pointer")
+	}
+	logDrain.ID = *response.Payload.ID
+	logDrain.DrainType = *response.Payload.DrainType
```

Comment:
Don't we need to guard all of these pointer deferences to avoid panics, not just the first? As it is, this code leaves us open to randomly panicing and crashing terraform.

Reply:
Here's an example of us guarding all the pointer accesses for reference: https://github.com/aptible/go-deploy/blob/810f7fb981bd26f9e197d9bf973f0dab3ae3968f/aptible/service.go#L30

Reply:
So this feels like overkill and is screaming for a more maintainable solution.  I found this: https://github.com/go-swagger/go-swagger/issues/1386

I think I'll use the helper functions found here: https://github.com/go-openapi/swag/blob/master/convert_types.go

Reply:
There's a bunch of those types of util packages floating around like for the [AWS SDK](https://github.com/aws/aws-sdk-go/blob/v1.39.2/aws/convert_types.go), [K8S](https://github.com/kubernetes/utils/blob/master/pointer/pointer.go), or [standalone packages](https://github.com/openlyinc/pointy). I don't feel strongly about them one way or the other. If you prefer to use a helper over the if statement, WFM. as long as the nils are checked!